### ISSUE TITLE
Refactor traits based on methods

### ISSUE THREAD

**Issue by bramtayl** [JuliaLang/julia#35095]
All traits which semantically define whether or not a method is expected, e.g. `HasShape`, `HasLength`, `HasEltype`, could be refactored into a single function:

`expect_method(::typeof(size), ...)`
`expect_method(::typeof(length), ...)`
`expect_method(::typeof(eltype), ...)`

which need only take two typed values: `True` and `False`. This could unify many existing trait interfaces, and allow them to be more easily combined. In addition, I propose

`expect_method(::typeof(getindex), ::Type{Container}, ::Type{Vararg{Int, Dimensions}})`
`expect_method(::typeof(setindex!), ::Type{Container}, ::Type{Element}, ::Type{Vararg{Int, Dimensions}})`

etc. for any method that is required by an Abstract interface.

**Comment by yuyichao** [comment]
> which need only take two typed values: True and False

It is still better to use value, not types. Working around inference limitation is strictly a problem for the caller of the function.

> This could unify many existing trait interfaces, and allow them to be more easily combined.

How does this allow things to be more easily combined? It seems that this is similar to say there should be only one function and all other functions should be implemented as method for this single function,

**Comment by bramtayl** [comment]
> It is still better to use value, not types

Happy to change the proposal to `true` and `false` if that makes things easier.

> similar to say there should be only one function

I think its more like saying there should be only one method table...but fair point. I can imagine a macro which would make writing these definitions easier. Something like:

`@add_trait size(something::MyType) = ...

**Comment by bramtayl** [comment]
Also, worth noting that I'm not proposing about adding it for every function, just every function that is required as part of an abstract interface.

**Comment by yuyichao** [comment]
> Also, worth noting that I'm not proposing about adding it for every function, just every function that is required as part of an abstract interface.

In that case, I assume you mean automatic, since an opt in behavior won't be any different from what's currently available, just with a diifferent syntax.

However, if it's automatic, it'll basically be asking for a inferable version of `hasmethod`. And in some sense this isn't something that can be done automatic anyway or you are limited to what the type system can express and won't be able to do anything more flexiable than that.

**Comment by bramtayl** [comment]
> an opt in behavior won't be any different from what's currently available, just with a diifferent syntax

yes, I am envisioning opt in behavior, and yes its just a syntax change, but syntax matters

**Comment by yuyichao** [comment]
> yes its just a syntax change, but syntax matters

Sure, I agree syntax does matter. However, then the question is the claim you made initially.

> and allow them to be more easily combined

which is what I asked in [the first reply above](https://github.com/JuliaLang/julia/issues/35095#issuecomment-598420374).

**Comment by bramtayl** [comment]
> How does this allow things to be more easily combined?

Because you can use logical operators on Bools (and typed bools provided the proper infrastructure). So you could write for example

`arraylike(T) = expect_method(size, T) & expect_method(eltype, T) & expect_method(getindex, T, Vararg{Int}) & expect_method(setindex!, T, eltype(T), Vararg{Int})`

**Comment by yuyichao** [comment]
> Because you can use logical operators on Bools (and typed bools provided the proper infrastructure).

You can already do that now.

```
julia> Base.IteratorEltype([]) isa Base.HasEltype && Base.IteratorSize([]) isa Base.HasShape
true
```

Again, if this is just syntax change that's fine and most of the time I'll have little opinion on what the best syntax should be. However, if this is just a syntax change, then it'll be hard to say this is more flexible. It could be shorter, it could be more convinient for certain usage, but it won't allow you to do what's impossible before or in cases that it couldn't be used before.

FWIW, this is also very closed to having `HasEltype(...) = true/false` Looking through the usage in base, it seems that such definition make a lot of existing usage harder to implement (they are probably still possible, just with longer and different syntax) so it seems that it's more likely to be a wash on the syntax complexity for simple usage.

**Comment by bramtayl** [comment]
> such definition make a lot of existing usage harder to implement

Why?

**Comment by yuyichao** [comment]
The return value carries more than two states, and some functions dispatches on particular combinations. This methods is requiring all functions to have only two states (i.e. one bit of info) so you'll need to pass on more traits in order to pass the same amount of information.

In another word, what is possible now by passing `IteratorSize`, you'll need to pass in sth like both `expect_method(length)` and `expect_method(size)`.

**Comment by bramtayl** [comment]
You could dispatch on `True` and `False`. I thought you said that dispatch was slow anyway?

**Comment by bramtayl** [comment]
> You'll need to pass in sth like both expect_method(length) and expect_method(size).

You could have a fallback method like:

`expect_method(length, T) = expect_method(size, T)`

**Comment by yuyichao** [comment]
> You could dispatch on True and False. I thought you said that dispatch was slow anyway?

Yes, and I never say it's slow. I say it's no faster than a branch and in this case the comparison is to other dispatch and there won't be much difference.

Dispatchig on true/false does have additional problem though since you can quickly run into boolean hell with more and more traits involved all having the same value set. And it seems that you are indeed most interested in complex cases.

> You could have a fallback method like:

The produce of the info isn't the problem. The consumer is. If you are implementing something that dispatches on `IteratorSize` in order to pick the best method,

Now you have `f(::<IteratorSize>, ...)`, with this you need `f(::<expect_method(length)>, ::<expect_method(size)>, ...)`. Of course each method probably only care about one value, but you have to pass both to make the dispatch work in general case.

**Comment by bramtayl** [comment]
> Now you have `f(::<IteratorSize>, ...)`, with this you need `f(::<expect_method(length)>, ::<expect_method(size)>, ...)`.

This seems like extra flexibility, not extra complexity, to me. Also, I think something more like:


```
if expect_method(size, x) isa True
    ...
elseif expect_method(length, x) isa True
    ...
else
    ...
end
```

would be better.

**Comment by yuyichao** [comment]
> This seems like extra flexibility,

It's equivalent functionality (i.e. not more flexible) and longer syntax (I'm only talking about syntax complexity in this case). I believe the latter objectively requires more argument. I'm making no argument on which one is better for this simple case, just that the complexity is higher.

```
if expect_method(size, T) isa True
    ...
elseif expect_method(length, T) isa True
    ...
else
    ...
end
```

No that's how you lost flexibility. It is of course a good way to implement pure logic on traits but you can't do that in general. This is exactly why we have dispatch at all and not just branching on type checks in user code. With that loop written, it is not impossible to extend the dispatch logic it creates which make it worse.

Also note that since you are dealing with only one type and one kind of info in this case you can easily exhaust all possible values never require this to be extended, which is exactly why this is a good way to implement pure/simple trait computation. However, as long as you have any other dispatch/logic in the branch, those logic are not bound to the branch logic and cannot be extended freely anymore. In another word, the branch is a single dispatch on one argument, which is strictly less flexible than multiple dispatch.

**Comment by bramtayl** [comment]
> With that branch written, it is now impossible to extend the dispatch logic`

What do you mean? Can't you define new methods of `expect_method` at any time?

> the branch is a single dispatch on one argument, which is strictly less flexible than multiple dispatch

To repeat, if you want dispatch, you can use `True` and `False`. It was you, not me, who insisted on using `true ` and `false`. There's no reason we can't add `Sometimes` or `Slow` or whatever values we want.

**Comment by yuyichao** [comment]
> What do you mean? Can't you define new methods of expect_method at any time?

Yes, but that's not part of the dispatch logic (on trait values).

You can have `f(x, y) = if expect...(length, x) g1(y) else g2(y) end` and now you've lost the ability to dispatch on both `x` and `y` in `g*`.

> To repeat, if you want dispatch, you can use True and False. 

Yes, which is why I was using dispatch in the previous response.

> It was you, not me, who insisted on using true and false.

`true` and `false` does not disable dispatch. You could even do `Val{true/false}`.

> There's no reason we can't add Sometimes or Slow or whatever values we want.

Well, if that's the case, then sure, the lack of information won't be a concern anymore. Though,

1. You mentioned that

    > which need only take two typed values: True and False.

    and I assumed it was a selling point. (and to some degree I agree, but I also see that it has drawbacks that I brought up)

2. I don't really see how you can return info for both `length` and `size` in this case. It's not like `Sometimes` `size` should equivalent to `length`. I mean, sure it can, and you can probably even return `HasLength` when you query for `size`. However, I don't really see how this fits in naturally while maintaining your desired property of easy computation/logic on them.

------

I also see why you might be confused about the relation between dispatch and branch. So to summarise and also mention why I brought up branch vs dispatch in different context.

1. Dispatch is strictly more flexible than branches. Branches (at least in call cases brought up in this discussion) is just a non-extensible dispatch on a subset of arguments. In most cases it's even single dispatch.

2. Exactly due to this inflexibility, branches, when they can be used (less flexible cases) are basically always no slower than dispatch.

3. For basically all the use cases you've explicitly brought up before (not including the one I mentioned in https://github.com/JuliaLang/julia/issues/35095#issuecomment-598488247). They belongs to usecases that can be covered with branches. That's why

    1. I mentioned you should be using branches with simple values in response to your usecases (i.e. they are simple, conceptually, enough that you can use branch)
    2. You can create syntax that fits the subset of usage very well, better than what we currently have.

4. That's also why I said that the syntax does not fit so well with other usage, when you need to use the full power of multiple dispatch. For the example I use tot show this, the point is to take advantage of multiple dispatch, which is why replacing it with branches will be less flexible. To be clear, using the branch still won't be slower, but it'll just be less flexible, for a case that actually matter now.

5. Now again, both syntax should be equivalent at some level. The one you propose is probably simpler in some cases but won' be simpler in general. This is not a direct reason to reject anything (though it doesn't make a good supporting argument either) of course since multiple syntax doing the same thing while for different cases is allowed. It'll be better if you can make it directly compatible with the current way (i.e. whatever syntax used now still works) or it'll just be an argument between two equivalent syntaxes neither of which is more flexible than the other. As I mentioned above, both syntax has advantage over the other in some cases and I'm not very interested in deciding which one is "better".

**Comment by bramtayl** [comment]
> if you can make it directly compatible with the current way

This seems pretty easy to do. What about defining a fallback method like

`expect_method(size, T) = IteratorSize(T) isa HasShape`

**Comment by bramtayl** [comment]
> the syntax does not fit so well with other usage, when you need to use the full power of multiple dispatch

Sure, I agree. There's definitely traits that don't make quite as straightforward sense using this framework, like `IsInfinite` (though `expect_method(last, x)` seems like it would work). I don't see why we can't keep those as is.

**Comment by yuyichao** [comment]
> `expect_method(size, T) = IteratorSize(T) isa HasShape`

The general issue with this kind of approach is to keep the two in sync. Basically,

1. What should the user use? `expect_method` or `InteratorSize`? The whole point about usecases above is so that the user can use both.
2. What should be the user define. For `@add_trait` to work they have to define `expect_method`. However, in this particular case, I believe `HasShape` carries additional information that can't be derived from `expect_method` so one still need to define `HasShape`. If this was no an issue for some other cases, it'll mainly be about backward compatibility. In either case, you can't make all existing code to work and have to port all the defining code over.

> There's definitely traits that don't make quite as straightforward sense using this framework, like IsInfinite (though expect_method(last, x) seems like it would work). I don't see why we can't keep those as is.

I saw you use "unify many" instead of "unify all" and I wasn't commenting on that. (Since it was a fair statement assuming everything else is accepted). I'm talking about how you use a particular trait. Are you using it for something as inflexiable/simple as a branch or are you taking advantage of multiple dispatch. This is not about syntnax and the particular trait in question doesn't matter...

**Comment by bramtayl** [comment]
>  `HasShape` carries additional information that can't be derived from `expect_method`

You mean `expect_method(::typeof(size), T)`? Like what?

> What should the user use?

`expect_method`, and `HasShape` would remain only for backwards compatibility

**Comment by yuyichao** [comment]
> You mean expect_method(::typeof(size), T)? Like what?

Dimension,

```
julia> Base.IteratorSize(Vector{Int})
Base.HasShape{1}()

julia> Base.IteratorSize(Matrix{Int})
Base.HasShape{2}()
```

>  HasShape would remain only for backwards compatibility

Well, the whole point of the direct compatibilty I mentioned above is to use `IteratorSize` since it carries more bits of information along with it's own types to avoid boolean hell. If it is not to be used then it's the other option I listed of comparing equivalent syntax that favors different usecases. (And again, I'm fine with either as of now. I'm just clarifying the claims and pointing out all the limitations that I can think of for consideration.)

**Comment by bramtayl** [comment]
Ah got. I think the number of dimensions is just `length(size(x))`, no?

`boolean hell`

What is boolean hell? Is there a three headed dog there?

**Comment by yuyichao** [comment]
> I think the number of dimensions is just length(size(x)), no?

I believe so.

> What is boolean hell? Is there a three headed dog there?

I was actually looking for a wiki article about it but couldn't find one. There's a few google hit so you can read about it there. I actually also have no idea where did I first about this term...

Anyway, it's about using boolean to pass way toto much information and you end up with functions that looks like `f(false, false, true, true, false, true, false, false, false, true)` and lost track of what each parameter is actually doing, either on the caller side or on the callee side. That's why having more states and unique types helps. There'll be fewer (type) parameters and they are going to be more easily distinguishable from each other.

Of course you can always write bad code like this without using `Bool`, `f(0, 0, 1, 1, 0, 1, 0, 0, 0, 1)` isn't much better than the `Bool` version. I personally believe `Bool` is singled out since it's about as little information one parameter can carry and you'll maximize the number of parameters you have this way, leading to the hell more easily.

**Comment by bramtayl** [comment]
> you can always write bad code

Boolean hell just sounds like bad code hell to me. Which means there's a `treehaddg` there.

> lost track of what each parameter is actually doing

Sounds a lot like the existing collect machinery in `array.jl`

**Comment by yuyichao** [comment]
> Boolean hell just sounds like bad code hell to me. Which means there's a treehaddg there.

Yes, but since reducing the amount of information each parameter carries strongly encourages and even necessitate such usage the design of the argument is part of the bad code here...

In fact, I think most of what I've read/seen/done about avoiding such a trap are exactly about replacing boolean flags with enums, in some sense the design of the argument type is almost all of it.

> Sounds a lot like the existing collect machinery in array.jl

I would't call it easy to understand but it's far from what I'm talking about. The parameters are mostly following a defined order (so when you don't care about it you can assume you cann jut pass them in the same order you get them). The parameters are mostly of different types so there's no question what a function that has `...(..., ::HasEltype, ...)` is implemented for and the compiler can check if the argument make sense. (You are not gonna get a match if you pass in `HasLength` here where as a `::True`/`::False` will happily accept a return value from a wrong trait...)

**Comment by bramtayl** [comment]
`::True/::False` will happily accept a return value from a wrong trait...)

That's fair, and why I think inline use like

```
if expect_method(size, T) isa True
    ...
elseif expect_method(length, T) isa True
    ...
else
    ...
end
```

would be best

**Comment by yuyichao** [comment]
> That's fair, and why I think inline use like

The two are not comparible. There are cases, i.e. externally extendable multitple dispatch, that just can't be expressed this way. I'm only talking about the case you do need to pass trait value around here.

**Comment by bramtayl** [comment]
> You can have `f(x, y) = if expect...(length, x) g1(y) else g2(y) end` and now you've lost the ability to dispatch on both x and y in g*.

I don't buy this argument. You can do

```
if expect_method(size, T) isa True
    if expect_method(eltype, T) isa True || expect_method(somethingelse, T)`
        ...
    else
        ...
    end
elseif expect_method(length, T) isa True
    ...
else
    ...
end
```

which is just as flexible as multiple dispatch

**Comment by yuyichao** [comment]
> which is just as flexible as multiple dispatch

There's very little to be argued here. If it was as flexible as dispatch we won't have the current dispatch system at all.

As I already mentioned above, if you have control over everything, of course you'll know all the dispatch path when you write the code. However, that's simply not the case. Your package will most likely implement some fallback methods that uses the information you know, the function will look like `f(x::T1, y::T2)`. However, if you only have this, it'll be impossible for someone to implement a specialization of this function. The specialization may want to catch `x::MySpecialType` and `expect_method(length, y)`. If all what you can use is the branch, you simply can't add such a dispatch rule after your original definition is already defined.

And to repeat I've said before, you need to be clear if what you want to define using this is a dispatch or not. In another word, are you simply defining functions that takes `Any` parameters or are you going to use the trait you defined to define methods that only operate on some type combinations while allowing everyone else to define method operating on other type combinations **at the same time and in the same way**. The branch you are talking about complement this but cannot be used to do this.

**Comment by bramtayl** [comment]
> allowing everyone else to define method operating on other type combinations

Aha! I finally understand. That would make sense for some traits, but the ones I'm discussing are **only** the ones that are semantically equivalent to checking whether or not a method is expected (which is a surprisingly large number). `HasEltype` is probably the best example, because it only has one alternative: `EltypeUnknown`.

**Comment by bramtayl** [comment]
And I think a key benefit that you've been ignoring is that if we had a similarly defined `expect_method` trait for `getindex`, we could index iterators safely. Which is the whole reason I started on this quest.

**Comment by yuyichao** [comment]
Even in this case, the user may still want to extend based on a combination of property you know and property you don't know. Like I have

```julia
function f(x, y)
    if expect_method(length, x)
        ....
    elseif expect_method(length, y)
        ....
    end
end
```

some other user might very well be interested in specializing this only for `expect_method(length, y) && x::Array`. And without `expect_method(length, y)` being exposed in the signature it is impossible to do this.

> And I think a key benefit that you've been ignoring is that if we had a similarly defined expect_method trait for getindex, we could index iterators safely. Which is the whole reason I started on this quest.

I did't ignore it but I did not talk about it since it's only about syntax, and I acknowedged that I'm not too interested in deciding which syntax is better.

I'll also acknowledge that you can define this for `getindex`, but that's also why I asked (and then initially assumed) if you want to do this automatically. Without automatically doing this, there's really no difference between this and a `HasIndex` trait other than the specific synax. Sure, we don't have that but there's no reason we can't add it. Of course you can use `@add_trait` to make this more automatic (againn syntax) but that's really not much different from defining a `@define_index` macro that does the same for `HasIndex` (again, just syntax)

**Comment by bramtayl** [comment]
The `just_syntax` thing is what bothers me the most. Syntax is what makes the difference between a programming language than anyone can learn and a programming language that puzzles nearly everyone.

**Comment by bramtayl** [comment]
Programmers need to get over the elitist "if you don't know where you are, you don't belong here" attitude and just put up the street signs. Everyone belongs here.

**Comment by AriMKatz** [comment]
I don't think @yuyichao was doing or intimating anything of that sort. It was more of stating a  preference for the kind of code discussions which are of personal interest.

**Comment by bramtayl** [comment]
This is an easy question to answer. @yuyichao do you believe that everyone belongs here?

**Comment by yuyichao** [comment]
> Syntax is what makes the difference between a programming language than anyone can learn and a programming language that puzzles nearly everyone.

Sure, which is why for the few point that I feel it actually matters I pointed out the likelyhood of boolean hell.

For all other cases that I said I don't care much about/don't want to talk about/not talking about syntax, I'm saying that because,

1. I'm asking for clarification about your claim that things are more flexible and I believe **this** is not about syntax. If two syntax can express the same thing, this won't be a proof that one is more flexible than the other. Or,
2. I already acknowledged early on that [your syntax is easier to write a branch for](https://github.com/JuliaLang/julia/issues/35095#issuecomment-598484262). I even repeated it a few times below. I don't think there's any doubt here and indeed none of the discussion is about this. Or
3. There are two alternative syntax and I don't think one is clearly better than the other.

Assuming your latest

> this "just syntax" thing

refers to https://github.com/JuliaLang/julia/issues/35095#issuecomment-598529484, it'll be a combination of 1 and 3. You claim that you could do sth using your method and I'm saying that other than syntax you could add sth similar now. So it's again all about syntax so I then gave a comparison between the two syntax as much as I can think of and concluded that I don't think one is better than the other so I don't believe it's a clear benefit. I'm not even disgarding syntax as an argument but giving my reason to think there's no clear better option here on the syntax level. If you think one syntax is clearly better than another you are very free to give your own reason here, much like how I gave mine.

> This is an easy question to answer

It's as easy as a loaded question can be. Just because a question is asked as a yes/no question does no mean it has a clear yes/no question

> if you don't know where you are, you don't belong here
> do you believe that everyone belongs here?

1. clearly not **everyone**. I can easily think of people that don't.
2. I don't really know what you want to get at. I clearly think you belong to the discussion. You bring this up so by default you must be here to explain what you actually want. Clearly no one else is in a better position than you to explain what you actually want. I wouldn't just ping random person and ask him to tell me why you want to do this. Also, if I don't think you belong to this discussion, I would have closed the issue a long time ago. I would not have give you detailed explaination for the part that is not clear about what you want.

**Comment by bramtayl** [comment]
Well, I appreciate your time and input in helping me refine my ideas.

**Comment by bramtayl** [comment]
> there's no clear better option here

The thing that excites me about this system of traits is not the simple examples I can come up with the top of my head, but the more complicated examples that are possible:

```
if (method_exists(f1, T) & method_exists(f2, T)) | (method_exists(f3, T) & method_exists(f4, T))
    ...
else
    ...
end
```

which is certainly possible to do with multiple dispatch, but not nearly as nice.

**Comment by yuyichao** [comment]
> which is certainly possible to do with multiple dispatch, but not nearly as nice.

Well, I guess another issue I want to point out is that there are many properties what you are proposing have and they are all sort of independent and should really be discussed independently.

For example, in this case, the only property you are using is that the return value are generic values. This is what makes it possible to do logic on return value from different traits. Unfortunately, this is exactly one of the non-syntax issue I brought up above, it strongly restricts what information is allowed from a trait, requires more traits to carry the same information and prevent type checking from catching errors.

It does not require introducing a generic mechanism for all methods (one indepedent property that I don't think you've ever explained about).

It can also be achieved via defining sth like `HasLength(x) = IteratorSize(x) isa Union{HasLength,HasShape}` without relying on a trait with limited return value.

**Comment by bramtayl** [comment]
> a generic mechanism for all methods (one indepedent property that I don't think you've ever explained about)

I'm not advocating for anything automatic. I think it would be nice if we also used `expect_method` as a way to formalize the API of abstract types (e.g. `AbstractSet`s require this, that and this other fundamental methods) but not required. As far as I know, these interfaces are not explicitly written out for any abstract type with few exceptions (e.g. AbstractArray)

**Comment by yuyichao** [comment]
Right. The question is why do you want to tie all of those up in the same function when they are inherently about different functions anyway.

**Comment by bramtayl** [comment]
I would be happy with any system which would allow you to check for the expectation of API required methods at compile time. Maybe you can come up with a better design?

**Comment by yuyichao** [comment]
> Maybe you can come up with a better design?

There's nothing to design. You can just add sth like a `Index` trait that tells you about the validity of `getindex`, `setindex!` together with what kind of indexing they support (linear, multidimentional, Cartesian, etc) in whichever way you want.

**Comment by bramtayl** [comment]
Adding a new trait for every API interface function would be adding a whole lot of new names to Base. Not wholly opposed, that still seems better than what we've got.

**Comment by yuyichao** [comment]
It's no more names than what's already there. It also means that you don't need to tie each trait to one method and can have one trait per interface where it really belongs to.

As an example, `hash` and `isequal` belongs to the same interface, no matter which method you pick some users will be surprised.

**Comment by bramtayl** [comment]
I don't understand. Wouldn't you have to add new `HasHash` and `HasIsEqual` traits?

**Comment by yuyichao** [comment]
No you only want one.

**Comment by bramtayl** [comment]
How could you tell if something has `hash` but not `isequal`?

**Comment by bramtayl** [comment]
Oh I understand your point. 

> no matter which method you pick some users will be surprised

If you can define `isequal` using `hash` generically, that `isequal` is **not** a required part of any API and thus you don't need a trait for it

**Comment by yuyichao** [comment]
> How could you tell if something has hash but not isequal?

That's not valid code. The two functions belongs to a single interface. You should have either both or none.

**Comment by bramtayl** [comment]
I don't understand. If you can have a generic function like

`isequal(x, y) = isequal(hash(x), hash(y))` then why is `isequal` a required part of any abstract interface?

**Comment by bramtayl** [comment]
That in fact might be a perfect example of a generic function:

```
function isequal(x, y)
    x isa HasHash || error("Can't hash $x")
    y isa HasHash || error("Can't hash $y")
    isequal(hash(x), hash(y))
end
```



```

**Comment by yuyichao** [comment]
> If you can have a generic function like
> `isequal(x, y) = isequal(hash(x), hash(y))`

Why do you say so? That's a one way property you want to maintain. It's not a generic implementation. You can't have that.

**Comment by yuyichao** [comment]
> That in fact might be a perfect example of a generic function:

No. Again, you are misinterpreting dispatch again. You can't implement dispatch with branch.

**Comment by bramtayl** [comment]
Why not?

**Comment by yuyichao** [comment]
> Why not?

Not sure which one you were asking.

If you meat,

> Why do you say so? That's a one way property you want to maintain. It's not a generic implementation. You can't have that.

Then it's because that's not the correct way to implement things. Sure that method will maintain some invariance but it'll never be what you want to implement so it doesn't serve as a reasonable fallback either.

If you meant,

> You can't implement dispatch with branch.

Well, I've explained many times above. Branch is not dispatch. You cannot make a second version of this function that uses the same info to cover a different part of the type space.

**Comment by bramtayl** [comment]
But if its generic then you don't need a second version of the function. That's the definition of generic

**Comment by bramtayl** [comment]
I meant the second, but for the record:

```
that's not the correct way to implement things
```

Isn't a good reason for anything

**Comment by yuyichao** [comment]
> But if its generic then you don't need a second version of the function. That's the definition of generic

Right. You need a second one so this is not generic.

> Isn't a good reason for anything

I don't know what you mean but just to be clear you cannot implement isequal for bigint this way.

**Comment by bramtayl** [comment]
> You need a second one so this is not generic.

Ok... so then `isequal` is required as part of the API, and we do indeed need an `HasIsEqualTrait`. You're contradicting yourself

**Comment by yuyichao** [comment]
I never say we don’t need one. I said we need a single one for both since

> hash and isequal belongs to the same interface,

What you said about not needing one in https://github.com/JuliaLang/julia/issues/35095#issuecomment-598801225 is what I was clarifying against.

**Comment by bramtayl** [comment]
> hash and isequal belongs to the same interface

This is the part I don't understand. What does this mean?

**Comment by yuyichao** [comment]
It means

> You should have either both or none.

**Comment by bramtayl** [comment]
Why?

**Comment by yuyichao** [comment]
Because the two must be consistent with each other

**Comment by bramtayl** [comment]
Why?

**Comment by timholy** [comment]
This is a long thread and I haven't followed it. But in case it's relevant, dispatch based on *availability* of a method doesn't necessarily tell the whole story. I might want to support `length(iter)` because it's occasionally useful, but if it's an `O(N)` algorithm rather than a `O(1)` algorithm it might change how other algorithms should dispatch.

**Comment by bramtayl** [comment]
Point taken. I mentioned having the option of having a `Slow` return value above for this reason

**Comment by yuyichao** [comment]
> Why

All of the previous questions are pretty reasonable but I’m really not sure if this is a troll question or not since it’s clearly documented..... That’s just how things is defined and you can find it in the doc.
https://docs.julialang.org/en/v1/base/base/#Base.hash

**Comment by yuyichao** [comment]
> Point taken. I mentioned having the option of having a Slow return value above for this reason

Again, this means that you cannot do logic on them easily anymore.

**Comment by bramtayl** [comment]
> you cannot do logic on them easily anymore

Yes you can. You could define `Exists(T) & Slow(T) = Slow(T)` for example

**Comment by bramtayl** [comment]
I have never written an API with `hash` or `isequal`, so I don't know how you're supposed to set them up, but however you do it, its important to have a way to figure out whether an abstract interface has the methods or not.

**Comment by yuyichao** [comment]
That’s more like min or max. What should !Slow be?

> whether an abstract interface has the methods or not.

The point is that there isn’t a single method for this. It’s a multiple method interface.

**Comment by bramtayl** [comment]
It's a currently non existent interface. How do I figure out if a `BigInt` has hash defined or not?

**Comment by yuyichao** [comment]
If this is useful enough one can be add, say calling it HasIsequalHash. Similar to indexing above, I’m using existing interface as examples as how traits for them need to be defined.

**Comment by bramtayl** [comment]
I can't imagine a scenario when it would not be useful to know about the existence of API required methods. Can you?

Regarding fast/slow, seems to me a good solution would be having a separate `fast_method_exists` function with the following defaults. Both functions could just return `True` or `False`

```
fast_method_exists(...) = False()
method_exists(...) = fast_method_exists(...)
```

**Comment by yuyichao** [comment]
> I can't imagine a scenario when it would not be useful to know about the existence of API required methods. Can you?

Yes I can. It's when the API has property that isn't about existance of the method (i.e. when that's not what you are interesed in). You can obviously always still provide more information that's not useful, but having to use a completely different mechanism to query for different information isn't nice.

Also, it's not even about figuring out the existence of "methods". It's about it being "methods" rather than a "method".

> seems to me a good solution would be having a separate fast_method_exists function with the following defaults. Both functions could just return True or False

See above, I agree you can always encode more information by just using more bits. However, that's not a very good solution as soon as you want to implement dispatch.

**Comment by bramtayl** [comment]
>  when the API has property that isn't about existance of the method

For example?

> you can always encode more information by just using more bits

Two bits seems pretty good to me

**Comment by yuyichao** [comment]
> For example?

The type parameter for `HasShape`.

> Two bits seems pretty good to me

Not sure what you are talking about because you don't have two bits of information. You only have one bit per trait. If you meant that you think using two bits is fine then that's not the problem either. The issue is you need twice (or more) as many argument.

**Comment by bramtayl** [comment]
> The type parameter for HasShape.

We've been through this. `HasShape` gives redundant information if you know something has a size; you can just use `length(size(x))`

> you need twice (or more) as many argument.

You don't need any arguments. You just need `if`

**Comment by yuyichao** [comment]
>  if you know something has a size; you can just use length(size(x))

Well, I thought you weren't suggesting to **actually** use it that way. For one, this force you to actualy use the value so you can't do things in the type. Also, this is exactly what I meant that,

> having to use a completely different mechanism to query for different information isn't nice.

> You don't need any arguments. You just need if

Again, you can't use branch for dispatch. And again, if you are just talking about something that makes writing complex branch easier then it's fine. If you are talking about replacing trait on all what it can be used for then no, you can't use branches to replace dispatch and the necessity of putting things through arguments.

**Comment by bramtayl** [comment]
> you are just talking about something that makes writing complex branch easier then it's fine

Yes, that's what I'm talking about

**Comment by yuyichao** [comment]
> Yes, that's what I'm talking about

In that case please update the issue since things like

> Refactor traits
and
> unify many existing trait interfaces

are very misleading when what you are offering aren't to replace all what those are used for.

**Comment by bramtayl** [comment]
I'm totally lost. I'm talking about refactoring of all traits that only check for the existence of a method as logical. This makes complex branching easier.

**Comment by yuyichao** [comment]
I'm saying that as of now, a lot of the traits are used for dispatch (they are type domain computations and if you don't have to do dispatch you would not put these in type domain to start with). What you propose makes these usage much harder and you are not targetting this usage. Because of that, you should not advertise this as replacing traits because you can't do that.

**Comment by bramtayl** [comment]
You can use True and False for dispatch

**Comment by bramtayl** [comment]
Listen, we're going around in circles, and even if you did happen to convince me I'm wrong, I wouldn't stop arguing because your elitist attitude towards programming and the way you treat people makes me angry.

**Comment by yuyichao** [comment]
Well, you've already accepted that you are just interesed in operation on these and not dispatch. I also thought you already realized all the issues if you use those generic values for dispatch.

For review, see,

https://github.com/JuliaLang/julia/issues/35095#issuecomment-598485664
https://github.com/JuliaLang/julia/issues/35095#issuecomment-598488247
https://github.com/JuliaLang/julia/issues/35095#issuecomment-598491122
https://github.com/JuliaLang/julia/issues/35095#issuecomment-598496903
https://github.com/JuliaLang/julia/issues/35095#issuecomment-598510647
https://github.com/JuliaLang/julia/issues/35095#issuecomment-598513057
https://github.com/JuliaLang/julia/issues/35095#issuecomment-598522694
https://github.com/JuliaLang/julia/issues/35095#issuecomment-598526570
https://github.com/JuliaLang/julia/issues/35095#issuecomment-598529484
https://github.com/JuliaLang/julia/issues/35095#issuecomment-598537066
https://github.com/JuliaLang/julia/issues/35095#issuecomment-598851527
https://github.com/JuliaLang/julia/issues/35095#issuecomment-598872323

To summarize, yes, you can use generic value for dispatch and that's what leads to boolean hell. What you are proposing is a very bad way for the current usage of trait (I'm specifically tallking about the property of returning generic value from traits, not any aspect of it)

**Comment by yuyichao** [comment]
> I wouldn't stop arguing

OK, I take it as you are intentially going in circle here and make me repeating my argument for the sole purpose of keeping the argument going.

I think that's a clear enough sign that this issue can be closed.

**Comment by bramtayl** [comment]
> you are just interesed in operation on these and not dispatch

You can use `True` and `False` for branching or for dispatch. Branching works just as well as dispatch because of constant propagation. You only need most two bits for any of the examples we've discussed, which is maybe your definition of boolean hell but not mine.

**Comment by yuyichao** [comment]
> You only need most two bits for any of the examples we've discussed, which is maybe your definition of boolean hell but not mine.

I did not say any of them **are** boolean hell, I'm saying that it's how you get there, because

> The issue is you need twice (or more) as many argument.

And again, I'm saying you are at least doubling the number of argument for dispatch here. It's not going to be too much a problem when you have only one argument, which is most of the cases you brought up. It's only an issue when you are doing **multiple** dispatch.

FWIW,

> You only need most two bits for any of the examples we've discussed

This is not true. In both https://github.com/JuliaLang/julia/issues/35095#issuecomment-598526570 and https://github.com/JuliaLang/julia/issues/35095#issuecomment-598529484, I was explicitly talking about cases where more than one arguments and two bits of information is needed.

**Comment by bramtayl** [comment]
Yes, there's a slight increase in number of arguments compared to the existing trait system (I'm saying slight because I think having to check the speed of more than one method would be very rare), but at the expense of a consistent interface which can be extended arbitrary without any new types

**Comment by yuyichao** [comment]
> (I'm saying slight because I think having to check the speed of more than one method would be very rare), 

I've talked about this above as well. The point is that you are dispatching on the properties available. It's likely that most method will only use few bit of info, but the fact that **some** may need it means that you have to pass it around.

**Comment by bramtayl** [comment]
Yes

**Comment by bramtayl** [comment]
Closed in favor of #35940
